### PR TITLE
[GISel] Legalize bitreverse with types smaller than 8 bits

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds support for lowering `bitreverse` with types smaller than 8 bits. It also fixes an existing assertion failure in `llvm::APInt::getSplat`: https://godbolt.org/z/7crs8xrcG

The lowering logic is copied from SDAG:
https://github.com/llvm/llvm-project/blob/2034f2fc8729bd4645ef7caa3c5c6efa284d2d3f/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L9384-L9398

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
Test the vector cases too? Can we get some end to end IR tests?

**Comment by dtcxzyw** [comment]
> Test the vector cases too?

IRTranslator haven't supported vector intrinsics yet :(
See https://godbolt.org/z/xTfGMfEr3

**Comment by arsenm** [comment]
> > Test the vector cases too?
> 
> IRTranslator haven't supported vector intrinsics yet :( See https://godbolt.org/z/xTfGMfEr3

Maybe something goes wrong for scalable vectors, but vector intrinsics work?

**Comment by dtcxzyw** [comment]
> > > Test the vector cases too?
> > 
> > 
> > IRTranslator haven't supported vector intrinsics yet :( See https://godbolt.org/z/xTfGMfEr3
> 
> Maybe something goes wrong for scalable vectors, but vector intrinsics work?

It works. But we don't support legalizing G_OR for fixed vector types :(

> LLVM ERROR: unable to legalize instruction: %55:_(<8 x s2>) = G_OR %48:_, %54:_ (in function: test_v8i2)

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 1]
```diff
@@ -7977,25 +7977,48 @@ LegalizerHelper::lowerBitreverse(MachineInstr &MI) {
   const LLT Ty = MRI.getType(Src);
   unsigned Size = Ty.getSizeInBits();
```
Comment (dtcxzyw) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7978]: It should be `Ty.getScalarSizeInBits()`.

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 2]
```diff
@@ -7975,6 +7975,8 @@ LegalizerHelper::LegalizeResult
 LegalizerHelper::lowerBitreverse(MachineInstr &MI) {
   auto [Dst, Src] = MI.getFirst2Regs();
   const LLT Ty = MRI.getType(Src);
+  // FIXME: It should be getScalarSizeInBits. Please fix this when vector
+  // support is complete.
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:7979]: should just change it now

Comment (dtcxzyw) [reply llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp]: Done (without vector tests).

**Comment by dtcxzyw** [comment]
@arsenm Can I add some vector tests with `-global-isel-abort=0` as we did in https://github.com/llvm/llvm-project/pull/93063?

**Comment by arsenm** [comment]
> @arsenm Can I add some vector tests with `-global-isel-abort=0` as we did in #93063?

Yes, the only thing with those is one day someone will need to go back and remove all the -global-isel-abort=0 flags

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 3]
```diff
@@ -7975,27 +7975,50 @@ LegalizerHelper::LegalizeResult
...
+    // 6|7
+    //    [(val & 0xAAAAAAAA) >> 1] & [(val & 0x55555555) << 1]
+    // -> [(val & 0xAAAAAAAA) >> 1] & [(val << 1) & 0xAAAAAAAA]
+    SwapN(1, Dst, MIRBuilder, Swap2, APInt::getSplat(Size, APInt(8, 0xAA)));
+  } else {
+    // Expand bitreverse for types smaller than 8 bits.
+    MachineInstrBuilder Tmp;
+    for (unsigned I = 0, J = Size - 1; I < Size; ++I, --J) {
+      MachineInstrBuilder Tmp2;
+      if (I < J)
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:8006]: Braces, and split buildConstant to temp variable? The line break is pretty ugly

File: llvm/test/CodeGen/RISCV/GlobalISel/bitmanip.ll [hunk 4]
```diff
@@ -0,0 +1,207 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv32 -global-isel -global-isel-abort=1 -verify-machineinstrs < %s 2>&1 | FileCheck %s --check-prefixes=RV32
+; RUN: llc -mtriple=riscv64 -global-isel -global-isel-abort=1 -verify-machineinstrs < %s 2>&1 | FileCheck %s --check-prefixes=RV64
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/bitmanip.ll:3]: Don't need -verify-machineinstrs

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Pull request merged**